include/mscclpp/gpu_utils.hpp

```diff
@@ -118,6 +171,25 @@ Memory safeAlloc(size_t nelem) {
   return Memory(ptr, Deleter());
 }
 
+template <class T, T*(alloc)(size_t, size_t), class Deleter, class Memory>
+Memory safeAlloc(size_t nelem, size_t gran) {
+  if ((nelem * sizeof(T)) % gran) {
+    throw Error("The request allocation size is not divisible by the required granularity:" +
```

Comment:
Granularity or alignment? I think `cuMemAddressReserve` doesn't require the size to be multiple of the alignment.

Reply:
According to: https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1g899d69a862bba36449789c64b430dc7c, The size of this allocation must be a multiple of the the value given via [cuMemGetAllocationGranularity](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1g30ee906c2cf66a0347b3dfec3d7eb31a) with the [CU_MEM_ALLOC_GRANULARITY_MINIMUM](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gg3a202e4d32ae296db1af7efe75ce365dc74872d07341bb1ac24ccc4a1c9c2f56) flag

Reply:
not entirely sure the difference between alignment and granularity.

Reply:
Alignment is about the starting address, e.g. range 0x08~0x0c granularity is less than 4 bytes but aligned to 8 bytes. But I now understand we need granularity here, thanks!